exa/core/container.py

```diff
@@ -474,6 +481,7 @@ def __getitem__(self, key):
         raise KeyError()
 
     def __init__(self, name=None, description=None, meta=None, **kwargs):
+        self.log.info('adding {} attrs'.format(len(kwargs)))
```

Comment:
The logger methods, `info`, `debug`, `error`, etc. have really powerful formatting cooked in. For example, in this case we could consider doing something like `self.log.info('adding attrs', extra=kwargs)`. I found this while looking at the `logging.debug` (Python standard library) documentation.

Reply:
In order to leverage the `extra` keyword argument, we would need to update the `Formatter`'s format string. In the current implementation, we rely on inheriting that from the parent log's handlers, rather than setting them explicitly. 

To use the `logging` API, we would need to add logic for attaching (and deduplicating) "custom" `Handler`s with appropriate context aware `Formatter`s to support various keyword arguments. Then in each logging context, we would always need to provide every key in the format string or else it will not log properly.

The docs basically mention these two caveats of annoying key requirements and specializing the `Formatter` and `Handler`, so in this case my vote is for keeping it simple. Plain-old python string formatting is already so flexible and convenient (and 3.6 f-strings save lives), and simple methods or lambdas can replace the otherwise repetitive stuff where appropriate in the class heirarchy.